File: compilers/go/codegen/bn254_scalar_domain_test.go
package codegen

import (
	"math/big"
	"testing"
)

// Scalar-domain tests for the two BN254 G1 double-and-add ladders,
// EmitBN254G1ScalarMul and emitG1ScalarMulNamed (the Groth16 one).
//
// Both build k' = k + 3r and then run 255 iterations from bit 254 down to 0,
// with the accumulator seeded at bit 255 rather than stepped. That is sound
// ONLY for k in [0, r-1], and nothing reduced k. The scalars are Groth16
// PUBLIC INPUTS -- caller-supplied -- so the domain is attacker-chosen.
//
// The interval argument, redone for BN254 rather than assumed from
// secp256k1 (the numbers are NOT the same -- see EmitBN254G1ScalarMul):
//
//	r.bit_length   = 254        (secp256k1: n is 256 bits)
//	3r.bit_length  = 256        (secp256k1: 3n is 258 bits)
//	3r    >= 2^255              so bit 255 is set for every k in [0, r-1]
//	4r-1   < 2^256              so no bit above 255 is ever set
//
// Hence: seed at bit 255, iterate 254..0. That is exactly what both ladders
// already do -- the offset and the iteration count were right. What was
// missing is the precondition they rest on.
//
// Out of domain, both halves of the bound break:
//
//	k >= 2^256 - 3r  (~2.2902 r)   k' sets bit 256, which the loop never
//	                               reads, so the ladder returns a DIFFERENT
//	                               multiple of P instead of failing.
//	k <= -r                        k' drops below 2^255, so the seeded bit
//	                               255 is a lie and the result is garbage.
//
// Separately, sweeping c_i = k' >> i over the whole domain shows the
// Jacobian mixed-add's exceptional cases (c_i = 0, 1, 2 mod r) are reachable
// at i = 0 ONLY, at k = 0 (accumulator = -P) and k = 2 (accumulator = P).
// k = 2 was already handled -- bn254BuildJacobianAddAffineInline branches on
// H == 0 at every step -- but k = 0 was not: H == 0 with R != 0 also took
// the doubling branch, returning -2P where the answer is the point at
// infinity. A public input of 0 is entirely ordinary in Groth16.

// bn254AllZeroPoint is this codegen's encoding of the point at infinity: the
// affine x||y blob cannot represent O, so the Jacobian-to-affine conversion's
// Fermat inverse turns Z = 0 into 64 zero bytes. Same convention as
// secp256k1 / P-256 / P-384. It is not on the curve (0 != 0 + 3), so
// EmitBN254G1OnCurve rejects it.
func bn254AllZeroPoint() (x, y *big.Int) {
	return big.NewInt(0), big.NewInt(0)
}

// TestBN254G1ScalarMul_ScalarDomain exercises EmitBN254G1ScalarMul outside
// [0, r-1] and at the two reachable degenerate steps.
func TestBN254G1ScalarMul_ScalarDomain(t *testing.T) {
	gx := big.NewInt(1)
	gy := big.NewInt(2)
	r := new(big.Int).Set(bn254CurveR)
	p := new(big.Int).Set(bn254FieldP)
	zx, zy := bn254AllZeroPoint()

	x3, y3 := bn254ComputeKG(t, 3)
	x5, y5 := bn254ComputeKG(t, 5)
	x7, y7 := bn254ComputeKG(t, 7)

	// -G = (gx, p - gy)
	negGy := new(big.Int).Sub(p, gy)

	mul := func(a int64, b *big.Int) *big.Int { return new(big.Int).Mul(big.NewInt(a), b) }
	add := func(a, b *big.Int) *big.Int { return new(big.Int).Add(a, b) }
	sub := func(a, b *big.Int) *big.Int { return new(big.Int).Sub(a, b) }

	cases := []struct {
		name       string
		k          *big.Int
		expX, expY *big.Int
		why        string
	}{
		{"k=0", big.NewInt(0), zx, zy,
			"0*G is O; the last step adds P to -P, which is H == 0 with R != 0"},
		{"k=r", r, zx, zy,
			"r = 0 mod r, so also O -- needs the reduce, k' = 4r would otherwise be in range and give -2G"},
		{"k=3r", mul(3, r), zx, zy,
			"k' = 6r sets bit 256, above the loop's top bit"},
		{"k=r+5", add(r, big.NewInt(5)), x5, y5,
			"in-range residue, out-of-range scalar"},
		{"k=3r+7", add(mul(3, r), big.NewInt(7)), x7, y7,
			"bit-256 overflow with a non-identity answer"},
		{"k=-1", big.NewInt(-1), gx, negGy,
			"(r-1)*G = -G; OP_MOD takes the sign of the dividend, so the reduce must normalise"},
		{"k=-2r", mul(-2, r), zx, zy,
			"k' = r has bit 255 CLEAR, so the seeded accumulator bit is wrong"},
		{"k=-r+3", add(sub(big.NewInt(0), r), big.NewInt(3)), x3, y3,
			"k' = 2r+3 is also below 2^255"},
	}

	for _, tc := range cases {
		tc := tc
		t.Run(tc.name, func(t *testing.T) {
			mulOps := gatherOps(EmitBN254G1ScalarMul)

			var ops []StackOp
			ops = append(ops, pushPoint(gx, gy))
			ops = append(ops, pushBigInt(tc.k))
			ops = append(ops, mulOps...)
			ops = append(ops, pushPoint(tc.expX, tc.expY))
			ops = append(ops, opcode("OP_EQUALVERIFY"))
			ops = append(ops, opcode("OP_1"))

			if err := buildAndExecute(t, ops); err != nil {
				t.Fatalf("EmitBN254G1ScalarMul(G, %s): %s: %v", tc.k, tc.why, err)
			}
		})
	}
}

// TestBN254G1ScalarMul_AllZeroPointIsNotOnCurve pins the O encoding: the
// scalar-mul tests above assert 64 zero bytes for k = 0 mod r, and that is
// only a safe answer because the documented on-curve gate rejects it.
func TestBN254G1ScalarMul_AllZeroPointIsNotOnCurve(t *testing.T) {
	zx, zy := bn254AllZeroPoint()

	var ops []StackOp
	ops = append(ops, pushPoint(zx, zy))
	ops = append(ops, gatherOps(EmitBN254G1OnCurve)...)
	ops = append(ops, opcode("OP_NOT"))

	if err := buildAndExecute(t, ops); err != nil {
		t.Fatalf("bn254G1OnCurve(all-zero point) should be false: %v", err)
	}
}

// TestBN254G1AffineAdd_ZeroDenominatorIsNotMaskedToInfinity guards the
// PRECISION of the infinity mask added for P + (-P)
// (TestBN254G1Add_PointAtInfinity): it must key on px == qx AND py != qy, and
// must NOT key on a zero slope denominator.
//
// bn254G1AffineAdd uses the unified slope
//
//	s = (px^2 + px*qx + qx^2) / (py + qy)
//
// whose denominator is py + qy, and BN254 has j-invariant 0 with p = 1 mod 3,
// so F_p contains a primitive cube root of unity w. For any curve point (x, y)
// the point (w*x, y) is also on the curve, hence Q = (w*x, -y) has py + qy == 0
// while Q != -P -- and the true sum P + Q is an ORDINARY point, not O. Masking
// a zero denominator to the all-zero blob would answer "point at infinity"
// there: plausible and wrong, the exact failure mode 03f50d48 introduced on the
// NIST curves and f16790a9 had to undo.
//
// SEPARATE, PRE-EXISTING DEFECT, deliberately NOT asserted here: this input
// does not merely fail closed. Both the numerator and the denominator vanish
// (w^2 + w + 1 == 0 makes px^2 + px*qx + qx^2 == 0), Fermat gives inv(0) = 0,
// so s = 0 and the result is (-(px + qx), -py) -- which satisfies
// y^2 = x^3 + 3, because (1 + w)^3 == -1. bn254G1Add answers the zero
// denominator with a WRONG but ON-CURVE point, so assert(bn254G1OnCurve(r))
// does not catch it. Fixing that means replacing the unified slope with the
// chord/tangent selection secp256k1 and the NIST curves use, which also moves
// the shared Groth16 MSM path; it is out of scope for the P == -Q mask.
func TestBN254G1AffineAdd_ZeroDenominatorIsNotMaskedToInfinity(t *testing.T) {
	p := new(big.Int).Set(bn254FieldP)
	gx := big.NewInt(1)
	gy := big.NewInt(2)
	negGy := new(big.Int).Sub(p, gy)

	// w = primitive cube root of unity mod p: b^((p-1)/3) for the first small
	// b that does not land on 1.
	exp := new(big.Int).Div(new(big.Int).Sub(p, big.NewInt(1)), big.NewInt(3))
	var w *big.Int
	for b := int64(2); b < 100; b++ {
		cand := new(big.Int).Exp(big.NewInt(b), exp, p)
		if cand.Cmp(big.NewInt(1)) != 0 {
			w = cand
			break
		}
	}
	if w == nil {
		t.Fatal("no primitive cube root of unity found mod p")
	}
	if new(big.Int).Exp(w, big.NewInt(3), p).Cmp(big.NewInt(1)) != 0 {
		t.Fatalf("w is not a cube root of unity: %s", w)
	}

	// Q = (w*gx, -gy): on the curve, distinct from -G, zero denominator.
	qx := new(big.Int).Mul(w, gx)
	qx.Mod(qx, p)
	if qx.Cmp(gx) == 0 {
		t.Fatal("w*gx == gx: the endomorphism image is not a distinct point")
	}
	if new(big.Int).Mod(new(big.Int).Add(gy, negGy), p).Sign() != 0 {
		t.Fatal("py + qy is not zero: the input does not exercise the denominator")
	}

	zx, zy := bn254AllZeroPoint()

	var ops []StackOp
	ops = append(ops, pushPoint(gx, gy))
	ops = append(ops, pushPoint(qx, negGy))
	ops = append(ops, gatherOps(EmitBN254G1Add)...)
	ops = append(ops, pushPoint(zx, zy))
	ops = append(ops, opcode("OP_EQUAL"))
	ops = append(ops, opcode("OP_NOT"))

	if err := buildAndExecute(t, ops); err != nil {
		t.Fatalf("bn254G1Add(G, (w*gx, -gy)) must NOT be masked to the all-zero point: %v", err)
	}
}

File: packages/runar-testing/src/__tests__/bn254-scalar-domain.test.ts
import { describe, it, expect } from 'vitest';
import { emitMethod, emitBn254G1ScalarMul } from 'runar-compiler';
import type { StackOp } from 'runar-ir-schema';
import { ScriptVM } from '../index.js';

/**
 * The DOMAIN of `bn254G1ScalarMul`: every scalar, not just [0, r−1].
 *
 * The ladder builds k′ = k + 3r, seeds the accumulator at bit 255 rather than
 * stepping it, and then iterates bits 254…0. That is sound ONLY while
 * 2^255 ≤ k′ < 2^256, i.e. while
 *
 *     2^255 − 3r ≤ k < 2^256 − 3r      (≈ −0.3549·r … ≈ 2.2902·r)
 *
 * BN254's numbers are its OWN — r is 254 bits here against 256 on secp256k1 —
 * so 3r is 256 bits, 3r ≥ 2^255 and 4r−1 < 2^256. Outside that window the
 * ladder does not fail: it silently applies a DIFFERENT multiplier,
 *
 *     m = 2^255 + ((k + 3r) mod 2^255)
 *
 * which is ≢ k (mod r). Two consequences, both live:
 *
 *   • k ≥ 2^256 − 3r sets bit 256, which the loop never reads.
 *   • k ≤ 2^255 − 3r drops k′ below 2^255, so the seeded top bit is a lie.
 *
 * and it is MALLEABLE in the scalar: k and k + 3r are the same group element
 * but produce different points, while k = 2^256 − 1 and k = r − 1 are
 * different group elements that produce the SAME point.
 *
 * The fix is the same one secp256k1 and the NIST curves already carry: reduce
 * k to [0, r−1] up front with ((k mod r) + r) mod r — OP_MOD takes the sign of
 * the DIVIDEND, so the `+ r, mod r` is what normalises negative scalars.
 *
 * The reduce alone is NOT sufficient, and this file pins both halves. Reducing
 * makes k ≡ 0 (mod r) reachable (k = r, k = 3r, k = −2r all land on 0), and at
 * k ≡ 0 the last ladder step is handed accumulator == −base. The mixed-add's
 * H == 0 test cannot tell −base from +base, so without the additional R == 0
 * test at the final step that case takes the doubling branch and returns −2P
 * where the answer is the point at infinity.
 *
 * Oracle: a from-scratch double-and-add over y² = x³ + 3 written from the curve
 * parameters alone, cross-checked in the first test against values produced by
 * gnark-crypto's `bn254.G1Affine.ScalarMultiplication`. Everything runs through
 * the real @bsv/sdk `Spend` interpreter via ScriptVM, not emitter-level logic.
 */

/** BN254 base field. */
const P = 21888242871839275222246405745257275088696311157297823662689037894645226208583n;
/** BN254 group order (254 bits). */
const R = 21888242871839275222246405745257275088548364400416034343698204186575808495617n;
/** G = (1, 2). */
const GX = 1n;
const GY = 2n;

type Aff = { x: bigint; y: bigint } | null;

const m = (v: bigint) => ((v % P) + P) % P;

function inv(a: bigint): bigint {
  let r = 1n;
  let b = m(a);
  let e = P - 2n;
  while (e > 0n) {
    if (e & 1n) r = (r * b) % P;
    b = (b * b) % P;
    e >>= 1n;
  }
  return r;
}

function dbl(A: Aff): Aff {
  if (A === null) return null;
  const s = m(m(3n * A.x * A.x) * inv(2n * A.y));
  const x = m(s * s - 2n * A.x);
  return { x, y: m(s * (A.x - x) - A.y) };
}

function add(A: Aff, B: Aff): Aff {
  if (A === null) return B;
  if (B === null) return A;
  if (A.x === B.x) return m(A.y + B.y) === 0n ? null : dbl(A);
  const s = m(m(B.y - A.y) * inv(m(B.x - A.x)));
  const x = m(s * s - A.x - B.x);
  return { x, y: m(s * (A.x - x) - A.y) };
}

function mul(k: bigint, A: Aff): Aff {
  let acc: Aff = null;
  let cur = A;
  let e = k;
  while (e > 0n) {
    if (e & 1n) acc = add(acc, cur);
    cur = dbl(cur);
    e >>= 1n;
  }
  return acc;
}

const hex64 = (v: bigint) => v.toString(16).padStart(64, '0');
const pointHex = (pt: NonNullable<Aff>) => hex64(pt.x) + hex64(pt.y);
const ZERO_POINT = '00'.repeat(64);
const G: NonNullable<Aff> = { x: GX, y: GY };
const G_HEX = pointHex(G);

const blob = (h: string) => Uint8Array.from(Buffer.from(h, 'hex'));

/** The value bn254G1ScalarMul(P, k) MUST take: (k mod r)·P, with 0 ↦ O. */
function expected(k: bigint): string {
  const red = ((k % R) + R) % R;
  return red === 0n ? ZERO_POINT : pointHex(mul(red, G)!);
}

/** Run the emitted ladder for (G, k) through the real interpreter. */
function scalarMul(pointHexIn: string, k: bigint): string {
  const ops: StackOp[] = [
    { op: 'push', value: blob(pointHexIn) } as StackOp,
    { op: 'push', value: k } as StackOp,
  ];
  emitBn254G1ScalarMul((o: StackOp) => ops.push(o));
  const { scriptHex } = emitMethod({ name: 't', ops } as never) as { scriptHex: string };
  const res = new ScriptVM().executeHex(scriptHex) as never as { stack: Uint8Array[] };
  return res.stack.length
    ? Buffer.from(res.stack[res.stack.length - 1]!).toString('hex')
    : '(empty)';
}

describe('bn254G1ScalarMul scalar domain', () => {
  // --- oracle agreement -----------------------------------------------------
  //
  // Values produced independently by gnark-crypto v0.14.0
  // (bn254.G1Affine.ScalarMultiplication over the standard generator).

  it('the from-scratch ladder agrees with gnark-crypto', () => {
    const GNARK: Record<string, [bigint, bigint]> = {
      '2': [
        1368015179489954701390400359078579693043519447331113978918064868415326638035n,
        9918110051302171585080402603319702774565515993150576347155970296011118125764n,
      ],
      '3': [
        3353031288059533942658390886683067124040920775575537747144343083137631628272n,
        19321533766552368860946552437480515441416830039777911637913418824951667761761n,
      ],
      '5': [
        10744596414106452074759370245733544594153395043370666422502510773307029471145n,
        848677436511517736191562425154572367705380862894644942948681172815252343932n,
      ],
      '7': [
        10415861484417082502655338383609494480414113902179649885744799961447382638712n,
        10196215078179488638353184030336251401353352596818396260819493263908881608606n,
      ],
    };
    for (const [k, [x, y]] of Object.entries(GNARK)) {
      expect(pointHex(mul(BigInt(k), G)!)).toBe(hex64(x) + hex64(y));
    }
    // (r−1)·G = −G = (1, p − 2), also checked against gnark.
    expect(pointHex(mul(R - 1n, G)!)).toBe(hex64(1n) + hex64(P - 2n));
  });

  // --- the below-r common case must not move --------------------------------

  for (const k of [1n, 2n, 3n, 5n, 7n, 12345n, R - 1n]) {
    it(`ScalarMul(G, ${k}) === ${k}·G (in-domain control)`, () => {
      expect(scalarMul(G_HEX, k)).toBe(expected(k));
    });
  }

  // --- above the upper bound: bit 256 falls off the end of the loop ---------
  //
  // k + 3r ≥ 2^256. Note that k + r and k + 2r do NOT break — 5r < 2^256 — so
  // the smallest multiple of r that exposes this is 3r, not r.

  for (const [label, k] of [
    ['3r', 3n * R],
    ['3r + 7', 3n * R + 7n],
    ['2^256 − 1', (1n << 256n) - 1n],
    ['2^300 + 12345', (1n << 300n) + 12345n],
  ] as Array<[string, bigint]>) {
    it(`ScalarMul(G, ${label}) === (k mod r)·G`, () => {
      expect(scalarMul(G_HEX, k)).toBe(expected(k));
    });
  }

  // --- below the lower bound: the seeded top bit is a lie -------------------

  for (const [label, k] of [
    ['−2r', -2n * R],
    ['−r + 3', -R + 3n],
    ['−(r + 1)', -(R + 1n)],
  ] as Array<[string, bigint]>) {
    it(`ScalarMul(G, ${label}) === (k mod r)·G`, () => {
      expect(scalarMul(G_HEX, k)).toBe(expected(k));
    });
  }

  // --- small negative scalars (inside the window, pinned for the reduce) ----

  for (const k of [-1n, -3n, -12345n]) {
    it(`ScalarMul(G, ${k}) === (r${k})·G`, () => {
      expect(scalarMul(G_HEX, k)).toBe(expected(k));
    });
  }

  // --- scalar malleability --------------------------------------------------
  //
  // These are the adversarial vectors: a Groth16 public input is caller-chosen,
  // so an attacker picks the representative of the residue class.

  it('k and k + 3r are the same group element and MUST give the same point', () => {
    const k = 5n;
    expect(scalarMul(G_HEX, k + 3n * R)).toBe(scalarMul(G_HEX, k));
  });

  it('k = 2^256 − 1 must NOT collide with k = r − 1', () => {
    // Unreduced, 2^256 − 1 applies the multiplier r − 1 exactly, so the two
    // distinct residues produce one point. That is the malleability.
    expect(scalarMul(G_HEX, (1n << 256n) - 1n)).not.toBe(scalarMul(G_HEX, R - 1n));
    expect(scalarMul(G_HEX, (1n << 256n) - 1n)).toBe(expected((1n << 256n) - 1n));
  });

  // --- k ≡ 0 (mod r): the point at infinity ---------------------------------
  //
  // Affine x‖y cannot encode O and `bn254FieldInv` is Fermat, so inv(0) = 0 and
  // the ladder yields the all-zero blob. These are the cases the last step's
  // strict H == 0 AND R == 0 test exists for: at k ≡ 0 the accumulator going
  // into the final add is exactly −base.

  for (const [label, k] of [
    ['0', 0n],
    ['r', R],
    ['2r', 2n * R],
    ['3r', 3n * R],
    ['−2r', -2n * R],
  ] as Array<[string, bigint]>) {
    it(`ScalarMul(G, ${label}) is the all-zero point (k ≡ 0 mod r)`, () => {
      expect(scalarMul(G_HEX, k)).toBe(ZERO_POINT);
    });
  }
});
